nilearn/regions/signal_extraction.py

```diff
@@ -334,6 +342,12 @@ def img_to_signals_labels(
         labels_index = {l: n for n, l in enumerate(labels)}
         for this_label in missing_labels:
             signals[:, labels_index[this_label]] = 0
+
+    # finding the new labels image
+    masked_atlas = Nifti1Image(labels_data.astype(np.int8), labels_img.affine)
+
+    if return_masked_atlas:
+        return signals, labels, masked_atlas
```

Comment:
No this is something that won't break the previous codes. it is more like an additional argument so we can return an additional output (`masked_atlas`) without breaking the previous codes. I tried to make the change in a way that doesn't need deprecation (with help of @jeromedockes). We also tried to do it in a way (using parentheses) that does not require changes to [_filter_and_extract](https://github.com/nilearn/nilearn/blob/65783298e962542a0f0653460803f6e67335547e/nilearn/maskers/base_masker.py#L19)

Reply:
If we can avoid conditional arguments, that's better in terms of bug tracking and code maintenance.

Reply:
It would be better to deprecate the old behavior. I don't see a good reason we'd want to keep both behaviors. Just like we added `keep_masked_labels` to deprecate old behavior, we can also do this with `return_masked_atlas`  in the same deprecation cycle.

Another option would be to extract the added functionality into a separate new function in `signal_extraction.py` like:

```python
def _get_masked_data(imgs, labels_img, mask_img, background_label=0, keep_masked_labels=False):
    """Get the masked image."""
    labels_img = _utils.check_niimg_3d(labels_img)

    imgs = _utils.check_niimg_4d(imgs)
    _, labels_data = _get_labels_data(
        imgs,
        labels_img,
        mask_img,
        background_label,
        keep_masked_labels=keep_masked_labels,
    )

    return Nifti1Image(labels_data.astype(np.int8), labels_img.affine)
```

and then from `nifti_labels_maskers.NiftiLabelsMasker.transform_single_imgs` maybe under the call for `_filter_andextract` call:

```python
        self.region_atlas_ = _get_masked_data(
            imgs,
            self._resampled_labels_img_,
            self._resampled_mask_img,
            background_label=self.background_label
        )
```

In this case, there would be no modification to labels masker `_ExtractionFunctor` or to `img_to_signals_labels`

Reply:
@ymzayek yes! another idea is sth like that. Me and @jeromedockes had another idea which was defining a new function like `_img_to_signals_labels_with_masked_atlas` and then call this new function in `img_to_signals_labels` but don't return the `masked_atlas` output. This way we don't need deprecation right?
Which one do you think makes more sense? 1) a new argument and new output or 2) a new function?

              def img_to_signals_labels(
                  imgs,
                  labels_img,
                  mask_img=None,
                  background_label=0,
                  order="F",
                  strategy="mean",
                  keep_masked_labels=True,
              ):
              
                  signals, labels, _ = _img_to_signals_labels_with_masked_atlas(...)
                  return signals, labels

Reply:
@mtorabi59 Yes that works too, it's along the lines of what I was thinking so let's go with your implementation of adding a new function. Thx !